R/list.data.R

```diff
@@ -0,0 +1,154 @@
+#'      function that will be used to load the data. Contains a \code{character(0)}
+#'      if no suitable reader was found.
+#' }
+#'
+#' * Note that some readers return more than one variable, usually with the
+#'   listed variablename as prefix. This is true for for example the
+#'   \code{xls.reader} and \code{xlsx.reader}.
+#'
+#' @export
+#'
+#' @seealso \code{\link{load.project}}, \code{\link{show.project}},
+#'    \code{\link{project.config}}
+#'
+#' @examples
+#' library('ProjectTemplate')
+#'
+#' \dontrun{list.data()}
+list.data <- function(override.config = NULL) {
+  config <- .load.config(override.config)
+  .list.data(config)
+}
+
+.list.data <- function(config) {
+  # Get list of variables in data/, always recursive to exclude cached
+  # variables from nested files
+  all.files <- list.files(path = 'data', recursive = TRUE, include.dirs = TRUE)
+  # Get list of variables according to configured recursive_loading, used as
+  # filtering variable later
+  data.files <- list.files(path = 'data', recursive = config$recursive_loading)
+
+  # Get variable name and reader from filenames
+  files.parsed <- .parse.extensions(all.files)
+  varnames <- files.parsed$varnames
+  readers <- files.parsed$readers
+
+  is_ignored <- grepl(.prepare.data.ignore.regex(config$data_ignore),
+                      all.files)
+  is_directory <- file.info(file.path('data', all.files))$isdir
+
+  is_cached <- .is.cached(varnames)
+  cache_only <- rep(FALSE, length(varnames))
+
+  # Build the final data.frame
+  df <- data.frame(filename = all.files,
+                   varname = varnames,
+                   is_ignored = is_ignored,
+                   is_directory = is_directory,
+                   is_cached = is_cached,
+                   cache_only = cache_only,
+                   reader = readers,
+                   stringsAsFactors = FALSE)
+
+  # Get list of variables in cache/
+  cached.vars <- .cached.variables()
+  # Exclude variables already found in data/
+  cached.vars <- setdiff(cached.vars, varnames)
+
+  filenames <- rep('', length(cached.vars))
+  is_ignored <- rep(FALSE, length(cached.vars))
+  is_directory <- rep(FALSE, length(cached.vars))
+  cache_only <- rep(TRUE, length(cached.vars))
+  readers <- rep('', length(cached.vars))
+
+  # .cached.variables returns all variables without checking validity, need to
+  # call .is.cached to perform this check
+  is_cached <- .is.cached(cached.vars)
+
+  df2 <- data.frame(filename = filenames,
+                    varname = cached.vars,
+                    is_ignored = is_ignored,
+                    is_directory = is_directory,
+                    is_cached = is_cached,
+                    cache_only = cache_only,
+                    reader = readers,
+                    row.names = NULL,
+                    stringsAsFactors = FALSE)
+
+  # Keep only lines with files that match the configured recursive_loading
```

Comment:
Initially all data is listed with `recursive = TRUE`, this stage keeps only those files that should be loaded with respect to `config$recursive_loading`. There were some edge cases when alternating `recursive_loading` where variables from nested folders were loaded from cache if loaded previously, even with `recursive_loading = FALSE`. I do agree this comment should be made a little clearer. Given this story, do you have a suggestion for improvement?